### DISCUSSION TITLE
Feedback from Bob WR7Q

### DISCUSSION THREAD

**Discussion by scotthibbs** [w7sst/MorseRunner#174] [Ideas]
I thought I would share the feedback we got from Bob WR7Q on a zoom call with Mike W7SST and Scott KD4SIR on 31Jan2023. Bob is a CWOps Academy teacher and has taught morse code for years. 

1. Responders in the ARRL DX exchange is more realistic in 1.68 than in 1.82 in the sequencing of the number. And he has noticed no numbers beyond 43. (I think this is because he probably doesn't do longer than 30 minutes and these numbers are a response based upon time. Further testing is needed.)  
_Addressed in #269_

2. Single call mode should not need to use RIT - These should be on frequency.
_Addressed in #257_ 

3. Farnsworth: effective speed and character speed should be different settings. CWOps Academy sends characters at 25wpm (minimum) but effective speed could be slower like 15wpm. Discussed if this could be set in the ini file, menu or GUI. 

4. MorseRunner - class setting so students could send him results. Discussed website. However, the high score table file (morserunner.lst that is created on contest finish) could be expanded and or renamed to be emailed to the cwops adviser. 
_Addressed in #255 and #258_

5. Regional Contesting - Contestors around the world have different responses. We discussed if this would be a setting in the program or if the contest specific data file would reflect this. 

6. Installation instructions need to be written to explain that a person needs to expand the zip file in a new directory so that all the files can be updated. (Using MR with an old ini can give odd results etc.)
_New releases will have better install instructions, also addressed in #256_ 

7. NAQP or other contests - We send our exchange automatically when we should listen for an R to make sure we got the call right before sending our exchange.

8. Font needs to be bigger. Log is in a different font than the program. Prefer Consolas or other slash zero font. 

9. Mike noted that the morse code timing might need to be adjusted. There could also be an issue with the trail end of the code being clipped. 

10. The audio recording feature should pop up a "save as" dialogue so that the same file ins't being written over. 
_Addressed in #259_ 

11. Ghosting was discussed. Ghosting is different for HST where it's a set time, and a random element for all the other contests. Ghosting could also be a "band conditions" feature underneath the lids selection. Or a slider setting.  

 Some of these I'll create an issue to resolve and the rest I will leave for discussion here.

**Comment by w7sst** [comment]
Mike also noted:

12. No leading '0' on two digit numbers. When using `Settings | NR Digits | N` (where N = 1-4), MR does not need to add leading '0' in front of a 2-digit number. When selecting `NR Digits`=2, a 2-digit random number should be generated. Adding a leading '0' is not necessary in Bob's, @WR7Q, opinion. We noticed this in CW WPX Contest. I can imagine that for true contesters, they would want this as I think N1MM generates a 3-digit number with leading zeros.
13. In a followup discussion with Scott, when running these learning contests, such as SST, the idea of disabling the DxStation's timeout feature so the station does not leave the QSO. This gives the user the chance to complete the QSO. Perhaps this feature is useful in Single-Call mode as well, but not Pile-Up. Scott also added:
> no timeout in CWT and SST contests. everyone is learning, so don't leave.
14. Mike suggested that the DxStation slows down his sending speed after the user sends `F7  ?` several times. This would give the user the chance to complete a difficult callsign, especially in Single-Call mode. Perhaps this can be a user-selectable option. In contests, it is common for the other station to slow down their speed if the user is having a hard time copying the exchange.

**Comment by WR7Q** [comment]
Mike,

Regarding #12, see replace 2 digits with 3 digits.

Thanks for your help and good meeting today.

I appreciate all you efforts on upgrading Morse Runner.

73 Bob WR7Q

Ps my hair cut looks great. hi

**Reply by w7sst** [reply]
> Regarding [#12](https://github.com/w7sst/MorseRunner/issues/12), see replace 2 digits with 3 digits.

@WR7Q, can you explain your comment above a little more relating to "replace 2 digits with 3 digits". I was saying that the contester may want the leading zero. It will also change by operator - some will send the leading zero, others will not. So we will have a random behavior so both behaviors are presented.

**Comment by WR7Q** [comment]
Mike,

Sorry to take so long to get back to you.  It has been a very long day.

When we were chatting about CQ WPX, we talked about 2 digits, but it really
is 3 digits.  Not
'01', but '001'. All the exchanges are typically 001 to 045.

73,
Bob WR7Q
CWA Admin
8018796566

**Comment by w7sst** [comment]
@WR7Q Bob also mentioned that they teach their CWT students to never send their report unless they have heard 'R' from the other station. Turns out that in the CWT Contest implementation, the DxStation does not report exchange received with an 'R' report before sending Name and Number. Issue #175 as added to fix this problem.

**Comment by WR7Q** [comment]
Thanks for the feedback.

73 Bob WR7Q

**Comment by w7sst** [comment]
@WR7Q @scotthibbs @ct7aup @K6OK 
Bob, WR7Q, mentioned in item 5 above the notion of Regional Contests, meaning that in different parts of the world, the user will mostly hear callsigns near them (i.e. same region).

I had an idea that for some contests, we can filter callsigns from the call history file using Continent as a filter. Using the user's current callsign we can determine which continent he is operating from. This using this information, we can filter callsign by comparing the DxCC entry for each call against the user's continent. This would give a European user mostly Europe contacts, etc. We could then load 10% of the other continent callsigns. We can implement this by contest as Bob suggested. I can also put a switch in the .INI file to disable or enable the feature. I think this will work great for something like CWT which Bob is current advocating for.

There has also been a suggestion of providing multiple call history files, one for the various major regions. While this will work, it does add complexity when trying to find the correct filename for the user. Using a DXCC-based solution, this might work okay for the short term.

Any thoughts on this idea?

Thank you.

**Comment by WR7Q** [comment]
Hi Mike,

I like the idea of a switch in the 'ini' file. This may be a bit much, but
I am wondering if you could use a number instead, to select the
Continent??  This could then allow the user to select which one he wanted
to practice with...

73,
Bob WR7Q

Ps I was a programmer for over 40 years...

CWA Admin

**Reply by w7sst** [reply]
Instead of a number in the .INI file, we could use the DXCC's continent abbreviation:  AF, AN, AS, EU, NA, OC, SA.

Perhaps something like the following to select all callsigns for a given continent:
```
[Callsign Filter]
SelectByContinent=<continent>[,<continent>...]
```
**Option 1**
We could consider adding a percentage number to select a subset of the calls in each continent grouping:
```
[Callsign Filter]
SelectByContinent=[<number>:]<continent>[,<continent>...];[<number>:]<continent>[,<continent>...]
```
The numbers specified above will select a percentage of the calls in each grouping. Using two groups with numbers 100 and 10 would select 100% of the first grouping and 10% of the calls in the second grouping. In the example below, we would select 100% of the calls from North America and 10% of the calls from Asia, Europe, and Africa combined.
```
[Callsign Filter]
SelectByContinent=100:NA;10:AS,EU,AF
```
While this idea is easy in a .INI file, it will be a little harder to present this in a future Settings Dialog box.

**Option 2** (My preference)
We could also present this in Continent:percent format. Something like:
```
[Callsign Filter]
SelectByContinent=<continent>[:<number>][;<continent>[:<number>]]...
```
The example below shows NA 100%, AF 10%, AS 10%, EU 10%
```
[Callsign Filter]
SelectByContinent=NA:100; AF:10; AS:10; EU:10
```
This form could be presented in a simple table in a future Settings Dialog box.

**Reply by K6OK** [reply]
I'm neutral on Option 1 or 2...  but whatever you do, please allow the user to override this and allow the user to supply their own call history files if they want to.  If the user supplies their own, then should be no manipulation of continents by percentages.

73 Jim K6OK

**Comment by WR7Q** [comment]
Hi Mike,

I like option 1 a lot.  Set the default for na 100 eu 10.  Then a dialog
box would be great down the road.

73,
Bob WR7Q
CWA Admin

**Comment by scotthibbs** [comment]
Further discussions with Bob WR7Q on Feb 9th 2024 yielded the following topics. (I removed those we already fixed or I already added
issues for)

**Ghosting:**

         He suggests holding somebody with f7 for at least three times then f5 for five times before ghosting. But many of his students want someone to ghost (hard or long call) so if after f5 is sent, if a F8 is hit, we want that person to ghost.

**Original Programming:** 

Bob really really wants the 1.68 pile up and single modes. He states ours is not the original programming because ours are based on the contest selected. For example, Field Day gives FD single calls. He suggests a way to make the run drop down all the original programming by creating an entry in the  contest drop down called "Original" (I suggest "Original 1.68 Program"). This allows the run button to be able to select the single call or pile up using the original programming. (This default NR digits to two for example). He would like as much as possible the original coding from 1.68 for the run button menu options when this contest (should we rename this mode?) is selected. 

**“Characters per Minute” added to settings menu below cw speed.**

CW Speed in the settings dropdown should be renamed “CW Farnsworth speed” (no change to gui that matches this setting)
Drop down for CPM should be:
18
20
25 (always as set as default)
Check box for “Do not override”
(btw, these speeds are used by CW Academy for their level of classes)

The "Do not override" option is needed. Normally if they select a Farnsworth speed (our normal cw speed) above 25, the CPM is matched. This option allows the user to select a higher Farnsworth speed but keep the CPM to something smaller.  
 
 **He suggested SOTA and POTA next and I think this would be absolutely brilliant!**

**Comment by K6OK** [comment]
@w7sst @scotthibbs @WR7Q 

> Bob really really wants the 1.68 pile up and single modes. He states ours is not the original programming because ours are based on the 
> contest selected. For example, Field Day gives FD single calls.

I don't quite understand the need for an "original program" setting.  In 1.68 if you select Single Calls you get the CQ WPX contest by default.  In 1.83 select CQ WPX as the active contest then Single Calls.  The result is the same.  How do we explain a 1.68 setting to new users?

> CW Speed in the settings dropdown should be renamed “CW Farnsworth speed” 

IMHO this only adds confusion for those of us interested in contest simulation where CW Speed is CW Speed.  One doesn't hear Farnsworth over the air in a contest (at least I haven't).

To be honest I'm a little concerned that adding a lot of options to fine tune the program between Contest Simulation mode and CWOps Training mode may make the program overly complicated for the end user to deal with.  Just a suggestion, but it might be worth considering forking the program into two editions, a Contest Simulator Edition and a CW Trainer Edition.  That will allow each to be tweaked for each use case and keeps the complexity of setup to a minimum (fewer menu choices) for both.

73 Jim K6OK

p.s. -- Adding POTA and SOTA is a great idea

**Comment by WR7Q** [comment]
Hi Jim,

Thanks for all the work you are doing on this.

I agree we want simplicity and not complications.  I also like your idea of
a CW Trainer Edition...

Here are some of my issues with your CQ WPX version.  It is significantly
different than the version in 1.68
Forget about Farnsworth, that was just a thought

I  really like the idea of replacing the GUI Run drop down with
a Start - Stop and even a Pause

1.68
- generally always gives you 3 tries at F7 and then another four or more if
you select F5
- if you correct a call it stays with you till you get it correct for quite
some time

1.84
- does not do two digits even if you select two;
  when you select two it should be the same on send and receive
- if you correct a call and then him F7, it immediately goes to a new call
a lot of times
- really do not need all the leading zeros in the NR field on both sending
and receiving...just wastes time

I would be more than happy to do a zoom discussion about any of this when
you have time...if you like?

73,
Bob WR7Q
CWA Admin
8018796566

**Comment by w7sst** [comment]
Hi Jim and Bob,
Version 1.84 includes changes for both the F5 and F7 keys. These changes were originally coded by Larry, F6FVY, back in 2018. They have been included in MRCE under issues #260, #264 and #262. If interested, you can see the linked code changes from these Issues.

I would suggest exercising the new behaviors using our latest build [v1.84-rc1](https://1drv.ms/u/s!AiN4lELeOz01q1bYfmH8YVW2-9KH?e=XWyb7J). If you have any questions, please reach out to Larry or myself. We want to verify that this new behavior is still working for our users. If we notice differences or concerns with the change, I many need to add a switch to the .ini file that will allow us to revert to the prior behavior.

Thank you,
Mike

@WR7Q @K6OK @f6fvy @scotthibbs

**Comment by w7sst** [comment]
@WR7Q 
Bob, thanks for the detailed description of observed differences between 1.68 and 1.84 for CQ WPX contest. I will look into these over the next few days and see what I can do about them. I will at least document the differences in behavior so we can decide what we should do next.

Tnx and 73, Mike

**Comment by WR7Q** [comment]
Hi Mike,

The F5 and F7 results were from 1.84-rc-1.

73,
Bob WR7Q
CWA Admin

**Comment by K6OK** [comment]
Mike, Bob --  I'm testing 1.84-rc1 right now.  Called CQ on WPX, got 2 stations calling, I hit F7.  Both replied after 4 F7's, then one went away, then after two more F7's the other left.  Had to call CQ again.  Then I did the repeated partial call with F5's and the station gave up after the 5th one.  In a real contest stations will ghost after a few F5's or F7's, so ghosting is a good simulation behavior.  So far I haven't seen the instant ghosting behavior that Bob reports.

2 digits -- in WPX, my serial number starts at T11.  Calling stations sometimes send single digit with two leading cut numbers.  Agree with Bob that calling stations should all have serial # > 10 if 2 digits selected.   The leading cut numbers seem to be set at three chars, emulating Cut Numbers = ON in N1MM.   Ideally the use of leading "O" or "T" or none at all should be randomized for callers to simulate what's on the air.  For the user's station ideally there should be an option to turn leading zeros on or off.

73 Jim K6OK

**Comment by WR7Q** [comment]
Hi Mike,

CQ WPX Single Call

Sorry, the F5/F7 process seems to be working just fine.

NR Digits --
1 = 1 thru 9 only
2 = 1 thru 9 only with a leading '0'
3 = 1 thru 99 with leading '0'
4 = 1 thru 999 with leading '0'

Issue -- when you space back to correct a call, then get it correct, you
get back an 'r' and it moves on, but the cursor still stays in the callsign
field instead of move to the NR field

Overall, I think it is working really great and I love seeing the wpm speed
worked.

Thanks.

73,
Bob WR7Q
CWA Admin

**Comment by f6fvy** [comment]
Hi Bob

> NR Digits --
> 1 = 1 thru 9 only
> 2 = 1 thru 9 only with a leading '0'
> 3 = 1 thru 99 with leading '0'
> 4 = 1 thru 999 with leading '0'

Are you sure ? I just tried 1.84-rc1 with NR setting to 1 and I got numbers with leading 0s (like 007) and even numbers above 010. I was running the CQ WPX contest in Pile-Up mode. Unless NR is restricted to specific contests / modes ? If so, that should be specified in the readme file (and the ideally NR settings menu enabled only in these cases).

**Comment by WR7Q** [comment]
Hi Laurent,

This is what I am suggesting they make to options for CQ WPX.  I don't
think that would work on some of the other contests...

73,
Bob WR7Q
CWA Admin

**Comment by f6fvy** [comment]
I'd rather suggest two settings. One for the number range and a different one for the leading zeros.

Ranges :
- 1 to 9
- 1 to 99
- 1 to 999
- 1 to 9999

Leading zeros:
- No leading zeros
- Leading zeros (for numbers < 100)
- Random (for numbers < 100)

The range 1 to 9999 with random LZ is IMHO the most realistic for the WPX contest.

**Comment by WR7Q** [comment]
I am great with that idea...

Thanks.

73,
Bob WR7Q
CWA Admin

**Comment by w7sst** [comment]
There is confusion regarding `NR Digits` menu pick. Last night, I added the following to the `Readme.txt` file. You can see my proposed change in [this review](https://github.com/w7sst/MorseRunner/pull/267). @scotthibbs will merge this into his `Readme.txt` file changes.

```
    NR Digits
      The number of digits the responding station will send.
        There are four possible settings:
          1 - generates random serial numbers starting near 1 and increasing as
	      the contest runs. Matches default behavior found in version 1.68.
          2 - 2 digits are sent, randomly generated between 1 and 99.
          3 - 3 digits are sent, randomly generated between 1 and 999.
          4 - 4 digits are sent, randomly generated between 1 and 9999.
```

The first choice, `NR Digits -> 1`, represents the original default behavior. The `NR Digits` option was original coded by @ct7aup back in Nov 2021 (version 1.71a).

Here is the code used to generate the serial number based on the selected option:
```
function TDxOperator.GetNR: integer;
Var
 n1: integer;
begin
  if NRDigits = 1 then
      Result := 1 + Round(Random * Tst.Minute * Skills)
  else begin
       n1 := trunc(power(10,NRDigits));
       n1 := n1-1;
       Result := Random(n1)+1;  // result is with range [1,99] (w/ NRDigits=2)
  end;
end;
```
As you can see, this is affecting only the generated integer number and does not control the leading zero behavior. `Tst.Minute` is the number of minutes elapsed since start of contest. Alex (original author) used this to simulate a growing serial number over the duration of the contest.

What is missing is using NR Digits to control leading zero behavior. At a minimum, we need to use `NR Digits` to control formatting the number into a string when being sent. For CQ WPX, this is used a hardwired `Format('%d%.3d', [RST, NR])`. This is forcing the `NR` to have exactly 3 digits, thus giving a 2 digit number a leading zero (e.g. 38 -> 038) and 1 digit numbers will have two leading zeros (e.g. 7 -> 007).

I like the idea of adding a `Settings -> Leading Zeros` menu option.

However, for now, I think I'll add code to format 2 digit numbers using `Format('%.2d', [NR])` to remove the leading zero on the 2 digit numbers. For serial numbers in the range 1-9, I think I'll leave them as having 2 leading zeros in the form of 'TT<n>'.

**Comment by w7sst** [comment]
We can also consider changing these menu picks to better describe these options to eliminate the confusion for the users. Perhaps something like:
- NR Digits > Normal
- NR Digits > 1 to 99
- NR Digits > 1 to 999
- NR Digits > 1 to 9999

**Comment by WR7Q** [comment]
Do we need to discuss leading zeros?

**Comment by K6OK** [comment]
> There is confusion regarding NR Digits menu pick.

There is... I can only speak for myself but with NR digits I'm looking for this behavior
  NR =3 means all serial numbers sent by callers will be 100 or greater NOT 1 to 999

The reason for this is if I'm practicing for WPX it's too easy to always reply to TT1, 002, etc.  Decoding and responding to 457, 149, or other 3 digit serials is a little bit more challenging.  I want to simulate the middle of the contest, not just the start of it.

But if the group thinks most will want the "NR Digits > 1 to 999" style I'm happy to defer.

73 Jim K6OK

@w7sst @WR7Q @f6fvy

**Comment by f6fvy** [comment]
> I want to simulate the middle of the contest, not just the start of it.

Good point. Ideally, the user should be able to define the number range he wants to hear.

**Comment by scotthibbs** [comment]
Great Idea @f6fvy ! 

So like:
NR Digits > Normal  (in the readme, I'd have to define normal. I don't know what normal is.)
NR Digits > 1 to 99
NR Digits > 1 to 999
NR Digits > 1 to 9999
NR Digits > Custom (this setting would look at the ini file for a range?)

Or should we make the user just set a range in the gui?

**Comment by K6OK** [comment]
Suggest consider use "random" instead of "normal."  Also "Serial NR" instead of "NR Digits" to be a little more precise of what's being set.

Serial NR -- Random, 1 to 9999     (use the word "random" instead of "normal")
Serial NR -- 1 to 99
Serial NR -- 100 to 999
Serial NR -- 1000 to 9999

73 Jim K6OK
@scotthibbs @f6fvy @w7sst @WR7Q

**Reply by w7sst** [reply]
I like Jim's suggestion regarding using Serial NR instead of NR Digits...
> Serial NR -- Random, 1 to 9999 (use the word "random" instead of "normal")
> Serial NR -- 1 to 99
> Serial NR -- 100 to 999
> Serial NR -- 1000 to 9999

Jim also mentioned NR=3 would generate serial numbers in the range of 100-999, NR=2 --> 10-99, etc.
> I can only speak for myself but with NR digits I'm looking for this behavior
> NR =3 means all serial numbers sent by callers will be 100 or greater NOT 1 to 999

I agree with this as well. I think Bob was saying the same thing. So how about the following:
```
Serial NR -- Random, 1 to 9999
Serial NR -- 1 to 99
Serial NR -- 100 to 999
Serial NR -- 1000 to 9999
```

All these choice are random, so perhaps...
```
Serial NR -- 1 to 9999 (Default)   // or perhaps: 1 to 9999, growing
Serial NR -- 1 to 99
Serial NR -- 100 to 999
Serial NR -- 1000 to 9999
```

Or perhaps...
```
Serial NR -- 1 to 9999 (increasing)
Serial NR -- 1 to 99 (random)
Serial NR -- 100 to 999 (random)
Serial NR -- 1000 to 9999 (random)
```
